pkg/metricsadapter/provider/custommetrics.go

```diff
@@ -3,33 +3,342 @@ package provider
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	metricsChanel := make(chan *custom_metrics.MetricValueList)
+	wg := sync.WaitGroup{}
+	for _, cluster := range clusters {
+		wg.Add(1)
+		go func(clusterName string) {
+			defer wg.Done()
+			metrics, err := c.getMetricByName(ctx, clusterName, name, info, metricSelector)
+			if err != nil {
+				klog.Warningf("query %s's %s metric from cluster %s failed, err: %+v", info.GroupResource.String(), info.Metric, clusterName, err)
+				return
+			}
+			metricsChanel <- metrics
+		}(cluster.Name)
+	}
+	go func() {
+		wg.Wait()
+		close(metricsChanel)
+	}()
+	for {
+		metrics, ok := <-metricsChanel
+		if !ok {
+			break
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
+	}
+	var metrics *custom_metrics.MetricValue
+	// TODO(chaunceyjiang) The MetricValue items need to be sorted.
+	for i := range metricValueList.Items {
+		if metrics == nil {
+			metrics = &metricValueList.Items[i]
+			continue
+		}
+		// metrics is unique in one cluster, but it may exist in multiple clusters.
+		// for this situation, we need to add the value of all clusters.
+		metrics.Value.Add(metricValueList.Items[i].Value)
+	}
+	if metrics == nil {
+		return nil, provider.NewMetricNotFoundError(info.GroupResource, info.Metric)
+	}
+	return metrics, nil
 }
 
 // GetMetricBySelector will query metrics by selector from member clusters and return the result
 func (c *CustomMetricsProvider) GetMetricBySelector(ctx context.Context, namespace string, selector labels.Selector, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValueList, error) {
-	return nil, fmt.Errorf("karmada-metrics-adapter still not implement it")
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	wg := sync.WaitGroup{}
+	metricsChanel := make(chan *custom_metrics.MetricValueList)
+	for _, cluster := range clusters {
+		wg.Add(1)
+		go func(clusterName string) {
+			defer wg.Done()
+			metrics, err := c.getMetricBySelector(ctx, clusterName, namespace, selector, info, metricSelector)
+			if err != nil {
+				klog.Warningf("query %s's %s metric from cluster %s failed", info.GroupResource.String(), info.Metric, clusterName)
+				return
+			}
+			metricsChanel <- metrics
+		}(cluster.Name)
+	}
+	go func() {
+		wg.Wait()
+		close(metricsChanel)
+	}()
+
+	for {
+		metrics, ok := <-metricsChanel
+		if !ok {
+			break
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
+	}
+	if len(metricValueList.Items) == 0 {
+		return nil, provider.NewMetricNotFoundError(info.GroupResource, info.Metric)
+	}
+	return metricValueList, nil
+}
+
+func (c *CustomMetricsProvider) getPreferredVersion(discoveryClient *discovery.DiscoveryClient) (schema.GroupVersion, error) {
+	apiGroups, err := discoveryClient.ServerGroups()
```

Comment:
Can we construct a set to record the API groups in `controller.go`, the set will be updated when the `cluster.spec` is updated.
Optimizing this will improve the query's speed.

Reply:
I don't think that's the case. If a user removes `custom metrics` from the member cluster, there won't be any changes in `cluster.spec`.